aip/0194.md

```diff
@@ -0,0 +1,237 @@
+dictionary defines idempotent as](https://www.lexico.com/en/definition/idempotent),
+
+> denoting an element of a set which is unchanged in value when multiplied or
+> otherwise operated on by itself.
+
+One interpretation of this definition is that an operation is idempotent if
+repeating the operation results in the same _response_.
+
+This definition allows for retrying `READ` and `UPDATE` requests, since
+repeated runs always result in the same state and response code. However, this
+definition precludes retrying `CREATE` calls, since `ALREADY_EXISTS` may be
+returned unexpectedly (as in the above example). It also precludes retrying
+`DELETE` calls, since `NOT_FOUND` may be returned unexpectedly.
+
+Another intepretation of this definition is that an operation is idempotent if
+repeating the operation results in the same _state_ (or, value).
+
+This second definition allows for retrying all CRUD operations, since the
+interpretation is only concerned with state, and repeating any CRUD operation
+always results in the same state.
+
+This document's retryability guidances follows the latter interpretation,
+which allows for retrying all CRUD operations. The reason for this is that
+there's much more value for users to retry more operations, with little downside.
+
+Consider the downside to retrying the `CREATE` and `DELETE` operations. For
+users it generally just results in an extra code check:
+
+```
+resp, err := client.Create("Foo)
+if err != nil {
+  if status.Convert(err).Code() == codes.ALREADY_EXISTS { // <- extra code check
+    // It already exists.
+    resp = client.Get("Foo")
+  } else {
+    // An actual error.
+    panic(err)
+  }
+}
+_ = resp // TODO: Use resp.
+```
+
+The above code is quite natural and exists in many codebases already: it is the
+easy answer to asynchrony within a system, and a cheaper/more performant way
+of writing code that doesn't know about state (the alternative is to always
+perform a `READ` before a `CREATE`/`DELETE`, which is superfluous).
+
+In comparison to the relatively small downside, there is a large advantage.
+Users will _almost always_ retry `INTERNAL` and `UNAVAILABLE`. By building it
+into a client library instead, users are saved from implementing retry loops,
+deadlines, exponential backoff, and jitter.
+
+Finally, consider that most users _will have to write the if statement either
+way_. After all, a user who implements retry logic themselves will _still_ run
+into the same problem (and solution) described above.
+
+#### Absolute correctness
+
+There is an argument that some user may need to be absolutely sure of the
+state of a system before retrying a non-`GET` request. Such a user is an
+argument in favour of only retrying `GET` requests; never retrying `UPDATE`,
+`CREATE`, or `DELETE`. However, this set of users likely has very niche
+requirements, and is dwarfed by the set of users for whom retrying is
+advantageous.
```

Comment:
This paragraph is an assumption, and it is not at all clear to me that it is a correct one.

- I intuitively disagree that retries that end in false errors are advantageous over not retrying.
- I intuitively disagree that the edge cases -- which become virtually impossible to debug or reason about -- are unimportant, or that an opt-out (which the user has no reason to find) is a sufficient remedy.

Furthermore, this paragraph fails the "you can always add and can not always remove" test. In other words, if my intuition turns out to be wrong, but we adopt it, we can update our guidance. The same is not true in the other direction.

In other words -- we are both guessing, but my guess is safer.

Therefore, absent contrary data, I assert that a tie should go to the position of not retrying these methods unless a `request_id` parameter is present. Additionally, we should steer APIs toward adopting AIP-154.

Reply:
> I intuitively disagree that retries that end in false errors are advantageous over not retrying.

Please describe your reasoning. I can't have a rationale discussion about abstract intuition. :)

> I intuitively disagree that the edge cases -- which become virtually impossible to debug or reason about -- are unimportant, or that an opt-out (which the user has no reason to find) is a sufficient remedy.

Ditto.

> In other words -- we are both guessing, but my guess is safer.

What is your definition of safe? Everything I've described has no negative result to user's state.

> Therefore, absent contrary data, I assert that a tie should go to the position of not retrying these methods

Your reasoning does not take into account "good defaults" or "customer happiness". I think the "safer" things is to provide good, reasonable defaults, where the definition of safety is "customer won't be tempted to use a different cloud": if using our cloud means writing a bunch of needless boilerplate whereas another cloud is Easy and Enjoyable, this becomes a point in favour of not using GCP.

Reply:
> I intuitively disagree that retries that end in false errors are advantageous over not retrying.

...because you are essentially serving the user a success disguised as an error, and with no clear debugging strategy. The user does not understand _why_ they got the error, and is unlikely to have written handling logic. `UNAVAILABLE` fundamentally means, "something might have gotten lost here, and you should go check it out".

This discussion is actually starting to mirror the discussion in #68 in a lot of ways, so it might be worth reading the discussion on that bug. What it boils down to is -- when something is unavailable, there is a fundamentally unknown state, and it is safer to say "we are in an unknown state, you need to look into it" than guessing.

> I intuitively disagree that the edge cases -- which become virtually impossible to debug or reason about -- are unimportant, or that an opt-out (which the user has no reason to find) is a sufficient remedy.

Lots of reasons here:

1. The two-actor situation could present a security vulnerability.
2. As described above, even the one-actor situation is really confusing.
3. The user has no reason to know that he or she was opted in, so has no reason to opt out until after the bad thing™ has already happened (which might involve data loss).

> What is your definition of safe? Everything I've described has no negative result to user's state.

But several of the situations I have described do.

> Your reasoning does not take into account "good defaults" or "customer happiness". I think the "safer" things is to provide good, reasonable defaults, where the definition of safety is "customer won't be tempted to use a different cloud": if using our cloud means writing a bunch of needless boilerplate whereas another cloud is Easy and Enjoyable, this becomes a point in favour of not using GCP.

It does take into account both good defaults and customer happiness. In the case of good defaults, I dispute what the good defaults are. In the case of customer happiness, which call would you rather take? The one of "why did I have to write this code to handle a 503 that you could have written for me?", or "why did you blow away my data?"

Reply:
> [user] has no reason to opt out until after the bad thing™ has already happened (which might involve data loss).

>  "why did you blow away my data?"

It seems as though you are primarily concerned with unexpected state changes (data loss via a deletion, etc). Could you describe to me a situation in which that could happen?

This document does very specifically state that a RPC in which unexpected state changes could happen as a result of retrying _should not_ be classified retryable.

Reply:
> > The two-actor situation could present a security vulnerability.
> > As described above, even the one-actor situation is really confusing.
> 
> You're alluding to situations I've not heard of and you've not described. Would you mind linking me to their descriptions, or writing about them?

I described it here: https://github.com/googleapis/aip/pull/270#discussion_r320844578

> I disagree. It means the request or the response was interrupted by a lack of connection. If we're talking about interpretations, I would say that a _much_ more prevalent interpretation is "retry me" - this is how LSTM, gRPC, most of Google backend service owners, and most of our customers tend to interpret it.

Right, but the gRPC team has said it is unsafe, as have several API producers (Storage, for example). I really wish it was the case that it was safe because I fundamentally agree with you that users would like it to be automatically retried.

Reply:
> I described it here: #270 (comment)

Thanks. Continuing that thread there.

> Right, but the gRPC team has said it is unsafe

They have said it is unsafe for operations which modify state. They are thumbs up on retrying requests which do not modify state. This document very specifically makes this distinction.